Fixes Array(Mutable/Immutable)BlockBuffer bugs

xcube16 opened the pull request:
This PR fixes three bugs that effect ArrayMutableBlockBuffer and or ArrayImmutableBlockBuffer.

Bug 1:
ArrayMutableBlockBuffer#getBlock() had a broken null check. (Optional#get() would crash before it got there)

Bug 2:
ArrayMutableBlockBuffer.BackingData#get() would return negative ids. Example: ByteBackingData would only support 128 local state ids.
http://i.imgur.com/v3WkeKy.png
(To the left is the original structure containing exactly 256 different block states. To the right is what was pasted form the schematic. Ignore the "modifying it in the world" chat message)

Bug 3:
ByteBackingData can not handle more than 256 state ids. This means that when ArrayMutableBlockBuffer#setBlock() adds a new local block state id past 255 it will wrap around!
http://i.imgur.com/CDsMIwW.png
(To the left is the original structure containing exactly 256 different block states. In the middle is the pasted schematic with sponge added **after** it was pasted. To the right is the pasted schematic with sponge added **before** it was pasted, causing it to wrap around and turn to bedrock.)

To fix this bug I added PackedBackingData to replace some of the old BackingData implementations. This is sort of like how net.minecraft.world.chunk.BlockStateContainer works with some with some improvements. Example: If there are only 10 local state ids, than PackedBackingData will only use 4 bits per block.

The only other BackingData implementation I left was CharBackingData as there are a lot of places that create an ArrayMutableBlockBuffer with a char[] array.
It might be a good idea to transition away from that.

http://i.imgur.com/KtM3Kak.png
(after all fixes)

xcube16 committed:
Fixed ArrayMutableBlockBuffer.BackingData#get() returning negative ids

Fixed broken null check in ArrayMutableBlockBuffer#getBlock()

xcube16 committed:
Fixed a bug in ArrayMutableBlockBuffer that caused local block state ids to wrap around.
 * Added PackedBackingData, this should also save memory
 * Removed ByteBackingData (No longer a valid BackingData)
 * Removed IntBackingData (unused and replaced by PackedBackingData)
 * Improved usage of ArrayMutableBlockBuffer to save memory

simon816 commented on src/main/java/org/spongepowered/common/util/gen/ArrayMutableBlockBuffer.java:
```diff
@@ -87,14 +85,13 @@ public BlockPalette getPalette() {
         checkRange(x, y, z);
         this.data.set(getIndex(x, y, z), this.palette.getOrAssign(block));
-        return true;
+        return true; // -254 111 250
```
I'm guessing this is a left over comment

xcube16 replied:
Oops. Ya, I was debugging and left that there.

Deamon5550 commented on src/main/java/org/spongepowered/common/util/gen/ArrayMutableBlockBuffer.java:
```diff
@@ -66,16 +70,10 @@ public ArrayMutableBlockBuffer(BlockPalette palette, Vector3i start, Vector3i si
     ArrayMutableBlockBuffer(BlockPalette palette, Vector3i start, Vector3i size, BackingData blocks) {
         super(start, size);
         this.palette = palette;
-        this.data = blocks.copyOf();
```
If you're going to change this then make it consistent with the immutable version (ie. remove the ctor in the immutable buffer that clones the data).

xcube16 replied:
Already did that.

Deamon5550 replied:
I'm talking about [this one](https://github.com/xcube16/SpongeCommon/blob/7621cc4a8028c385830ecfcd8a10350434c41e10/src/main/java/org/spongepowered/common/util/gen/ArrayImmutableBlockBuffer.java#L53), I'm not sure if its even used but any usages of it should be changed for consistency.

xcube16 replied:
O that one... There is only one usage, and thats [here](https://github.com/xcube16/SpongeCommon/blob/7621cc4a8028c385830ecfcd8a10350434c41e10/src/main/java/org/spongepowered/common/util/gen/ArrayMutableBlockBuffer.java#L132).
I like to think about it like new String(char[]), string being immutable, makes a copy of the char[] array so it cant be mutated form the outside.
However its ok if ArrayMutableBlockBuffer is changed from the outside using the char[].

This does get me thinking though... ArrayImmutableBlockBuffer() takes a palette argument. What if BlockPalette#remove() was called after ArrayImmutableBlockBuffer() was created. (Maybe this can be fixed in another PR)

Deamon5550 replied:
mm yeah we might need an immutable palette...

xcube16 replied:
Or we could remove the BlockPalette's remove() method... I don't like that method anyway, supporting it is why BimapPalette has a BitSet. (nothing in SpongeForge or any of its subprojects use it as far as I know, so its only there for the plugins)

Deamon5550 replied:
Yes  that is an option and it certainly would simplify things. It also presents problems even for mutable volumes.

Deamon5550 commented on src/main/java/org/spongepowered/common/util/gen/ArrayMutableBlockBuffer.java:
```diff
@@ -39,25 +40,28 @@
-        this.palette = palette;
-        this.data = type.create(size);
+    public ArrayMutableBlockBuffer(Vector3i start, Vector3i size) {
+        this(size.getX() * size.getY() * size.getZ() > LOCAL_PALETTE_THRESHOLD ?
```
This check seems backwards but what is your thinking?

xcube16 replied:
If there is a small number of blocks, say 20 blocks for example, using the global state ids is a good idea as it does not need to allocate its own local palette.
You get more savings from using a local palette if there is a larger area as it makes up for the cost of having its own BimapPalette.

Deamon5550 replied:
sure until the number of bits exceeds a threshold and the size of the palette starts being too much overhead compared to just using chars. MC only bit packs the chunks if the number of bits is 8 or less for example.

xcube16 replied:
net.minecraft.world.chunk.BlockStateContainer#setBits() does have some special cases for optimizing its palette. But it still uses a BitArray (bit packing) for storage.

Should I make it so that ArrayMutableBlockBuffer can transition to a global palette when the number of different states gets heigh enough?

Deamon5550 replied:
yeah that would be good I think

xcube16 replied:
This one is harder than I expected... Any ideas? I dont want it to be as messy as BlockStateContainer.

Deamon5550 replied:
should be as simple as iterating over the volume and translating the ids from the local palette to the global palette eh? Its more or less the same operation as your `remax` except with an added translation step.

xcube16 replied:
Right now, remax() does not need to deal with the palettes and stuff... It could use a callback and see if it should translate the ids also... But there is also a set() in flight that will also need to be translated at that point.
The way Mojang handles this is by putting a callback in the palette. (idk if I like that)

Another way to think of the problem... The local palette has gotten so full that it is almost as big as the global palette making it no longer worth having a local palette. This problem is not specific to the block buffer, but can happen to any local palette. :/
There has to be a better way to design all this, so its clean to implement.

xcube16 replied:
I think I found a solution.

Deamon5550 commented on src/main/java/org/spongepowered/common/util/gen/ArrayMutableBlockBuffer.java:
```diff
@@ -39,25 +40,28 @@
-        this.data = type.create(size);
+    public ArrayMutableBlockBuffer(Vector3i start, Vector3i size) {
+        this(size.getX() * size.getY() * size.getZ() > LOCAL_PALETTE_THRESHOLD ?
+                new BimapPalette() : GlobalPalette.instance, start, size);
```
If you're going to create a new BimapPalette then you need to make sure to add air to it as id 0. (I think this is the real cause of your first bug).

xcube16 replied:
I know its not the cause of bug 1. But you make a good point, I did not think about the fact that the buffer defaults to being filled with air blocks.
(In the images of my test plugin, setBlock() was called on all block postions, so there where no *undefined* blocks)

Deamon5550 commented on src/main/java/org/spongepowered/common/util/gen/ArrayMutableBlockBuffer.java:
```diff
@@ -39,25 +40,28 @@
-    public ArrayMutableBlockBuffer(BlockPalette palette, Vector3i start, Vector3i size, byte[] blocks) {
+    public ArrayMutableBlockBuffer(BlockPalette palette, Vector3i start, Vector3i size) {
         super(start, size);
         this.palette = palette;
```
Likewise here you need to make sure that that we're setting the contents of the backing data to the id for air in the incoming palette (which may not be zero).

xcube16 replied:
Fixed. (this should also fix ArrayMutableBlockBuffer(start, size) as it uses ArrayMutableBlockBuffer(palette, start, size))

Deamon5550 reviewed (changes requested):
some changes needed

xcube16 committed:
ArrayMutableBlockBuffer defaults to air, not undefined (thanks Deamon5550 for pointing that out)

xcube16 commented on src/main/java/org/spongepowered/common/util/gen/ArrayMutableBlockBuffer.java:
```diff
@@ -135,140 +132,167 @@ public ImmutableBlockVolume getImmutableBlockCopy() {
         @Override
         public int get(int index) {
-            return this.data[index];
+            return this.data[index] & 0xFFFF;
```
I think chars are unsigned, so I would not need the & 0xFFFF?

Deamon5550 replied:
yeah its not needed.

xcube16 committed:
Removed PackedBackingData#remax(). ArrayMutableBlockBuffer can now transition
to a global palette automaticly when it is more optimal.

Small formatting fixes

xcube16 committed:
Renamed LOCAL_PALETTE_THRESHOLD to SMALL_AREA_THRESHOLD and added javadoc so its easer to understand

xcube16 committed:
Added tests for ArrayMutableBlockBuffer (lots of hacks though)

Deamon5550 commented:
568f6ea4cddd55850f391e5228792bebad65960f

Deamon5550 commented:
I ended up not merging your unit tests due to the amount of hacking around the other elements of gamestate, we'd like to do more unit tests but they will be done in a more controlled stable fashion.
